land_grab_2/stl_dataset/step_2_5/get_cessions.py

```diff
@@ -0,0 +1,278 @@
+
+logging.basicConfig(level=logging.INFO)
+log = logging.getLogger(__name__)
+
+
+def area_join(
+    source_df: gpd.GeoDataFrame, target_df: gpd.GeoDataFrame, variables: List[str]
+) -> gpd.GeoDataFrame:
+    """
+    Join variables from source_df based on the largest intersection with feat-
+    ures from target_df. In case of a tie, pick the first joined record.
+    Implementation adapted from tobler:
+    https://pysal.org/tobler/_modules/tobler/area_weighted/area_join.html#area_join
+
+    Arguments:
+    source_df -- the GeoDataFrame containing source values
+    target_df -- the GeoDataFrame containing target values
+    variables -- string or list-like column(s) in source_df dataframe to join to
+                 target_df
+
+    Returns:
+    geopandas.GeoDataFrame -- target_df GeoDataFrame with joined variables as
+                              additional columns
+
+    """
+    for v in variables:
+        if v in target_df.columns:
+            raise ValueError(f"Column '{v}' already present in target_df.")
+
+    target_df = target_df.copy()
+    target_ix, source_ix = source_df.sindex.query(
+        target_df.geometry, predicate="intersects"
+    )
+    areas = (
+        target_df.geometry.values[target_ix]
+        .intersection(source_df.geometry.values[source_ix])
+        .area
+    )
+
+    main = []
+    for i in range(len(target_df)):
+        mask = target_ix == i
+        if np.any(mask):
+            main.append(source_ix[mask][np.argmax(areas[mask])])
+        else:
+            main.append(np.nan)
+
+    main = np.array(main, dtype=float)
+    mask = ~np.isnan(main)
+
+    for v in variables:
+        arr = np.empty(len(main), dtype=object)
+        arr[mask] = source_df[v].values[main[mask].astype(int)]
+        try:
+            arr = arr.astype(source_df[v].dtype)
+        except TypeError:
+            warnings.warn(
+                f"Cannot preserve dtype of '{v}'. Falling back to `dtype=object`.",
+            )
+        target_df[v] = arr
+
+    return target_df
+
+
+def join_counties_to_parcels(
+    parcels_gdf: gpd.GeoDataFrame, counties_gdf: gpd.GeoDataFrame
+) -> gpd.GeoDataFrame:
+    """
+    Join county name to parcels based on the largest spatial intersection.
+
+    Arguments:
+    parcels_gdf -- GeoDataFrame containing parcel geometries
+    counties_gdf -- GeoDataFrame containing county geometries
+
+    Returns:
+    GeoDataFrame -- STL parcels with county names joined as a new column
+    """
+    # Reproject counties_gdf to match parcels_gdf.
+    counties_gdf = counties_gdf.to_crs(parcels_gdf.crs)
+
+    out_gdf = area_join(
+        counties_gdf,
+        parcels_gdf,
+        variables=["NAME"],
+    )
+
+    # Drop the existing county column—we'll use the result of this join instead.
+    out_gdf = out_gdf.drop(columns=["county"])
+    out_gdf = out_gdf.rename(columns={"NAME": "county"})
+
+    return out_gdf
+
+
+def join_cessions_to_parcels(
```

Comment:
So one quick question here. In the doc, Maria notes the equivalent QGIS operation for this step would be **Join attributes by location** → _Take attributes of the feature with the largest overlap only - one to many_. However, that join parameter doesn't exist. If you look at [the docs](https://docs.qgis.org/3.34/en/docs/user_manual/processing_algs/qgis/vectorgeneral.html#id52), you can only choose from the following:

> The type of the final joined layer. One of:
0 — Create separate feature for each matching feature (one-to-many)
1 — Take attributes of the first matching feature only (one-to-one)
2 — Take attributes of the feature with largest overlap only (one-to-one)

I think she's actually after a `one-to-many` join here (all parcel-cession combinations are returned), which is what this code implements. However, the ordering of cessions in our later aggregation (i.e., which cession gets marked as `cession_num_01` vs. `cession_num_02`) is not done by area, so let me know if that ends up being important and we can try to work around it.

Reply:
Somewhat annoying, but I think it's fine. Definitely not worth dwelling on, especially given the structure of the downstream analyses.